boards/earlgrey-nexysvideo/src/main.rs

```diff
@@ -322,6 +339,53 @@ pub unsafe fn reset_handler() {
         i2c_master,
     };
 
+    // This is PMP support for kernel regions
+    // PMP does not allow a deny by default option, so all regions not marked
+    // with the below commands will have full access.
+    // This is still a useful implementation as it can be used to limit the
+    // kernels access, for example removing execute permission from regions
+    // we don't need to execute from and removing write permissions from
+    // executable reions.
+    let mut mpu_config = rv32i::pmp::PMPConfig::default();
+    // The kernel stack
+    chip.pmp
+        .allocate_kernel_region(
+            &_sstack as *const u8,
+            &_estack as *const u8 as usize - &_sstack as *const u8 as usize,
+            mpu::Permissions::ReadWriteOnly,
+            &mut mpu_config,
+        )
+        .unwrap();
+    // The kernel text
+    chip.pmp
+        .allocate_kernel_region(
+            &_stext as *const u8,
+            &_etext as *const u8 as usize - &_stext as *const u8 as usize,
+            mpu::Permissions::ReadExecuteOnly,
+            &mut mpu_config,
+        )
+        .unwrap();
+    // The kernel relocate data
+    chip.pmp
+        .allocate_kernel_region(
+            &_srelocate as *const u8,
+            &_erelocate as *const u8 as usize - &_srelocate as *const u8 as usize,
+            mpu::Permissions::ReadWriteOnly,
+            &mut mpu_config,
+        )
+        .unwrap();
+    // The kernel BSS
+    chip.pmp
+        .allocate_kernel_region(
+            &_szero as *const u8,
+            &_ezero as *const u8 as usize - &_szero as *const u8 as usize,
+            mpu::Permissions::ReadWriteOnly,
+            &mut mpu_config,
+        )
+        .unwrap();
+
+    chip.pmp.enable_kernel_mpu(&mut mpu_config);
+
```

Comment:
That's a good question. It doesn't really matter too much.

I feel that it fits in `boards/` as that is where other similar operations are. For example the stack size is allocated in boards, so I feel that stack protection should also be in boards.

We also only want to enable this after debug is setup (in case something goes wrong), so it makes sense in boards.

Reply:
The board can still call it, and decide when (or if) to call it, just the code would exist in a more central location.

Reply:
I have bad memories of duplicated code in main.rs, that's all. Components helped a lot, but I'm wary of setting things up so that code will get copied, and then changes require updating all RISC-V boards (which I do not recommend :)).

This is also similar to load_processes, which is only stored in process.rs for convenience, but is most definitely a board-level function/task. It's just copying it to every main.rs wasn't scalable.

Reply:
It would still required changes in all RISC-V if it's in chips.

I still think here is a good place. Maybe arch would work though, then it can just be called from all the boards.

Reply:
Ideally, this would not even be architecture specific, and tock could by default limit access on all platforms in the same way that this does for OT. But I think it is fine to leave it here for now and worry about code duplication once it actually exists.